detecter les anomalies liees aux données du personnel de maniere globale 
- PER : personnel Enseignat 
- PATS : Personnel administratif

In [61]:
import pandas as pd 
import numpy as np 

lire les données

In [62]:
data = pd.read_excel('C:/Users/HP/Desktop/mon_stage_senegal/data/Personnel_2000_2024/Liste_agent_des_archive 2000.xlsx') 

In [63]:
data.head()

ANNEE DATE DE PRISE EN COMPTE  NUMERO EMPLPOYE DATE RECRUTEMENT  \
0   2000              2000-12-31              784       1983-10-01   
1   2000              2000-12-31              927       1992-02-01   
2   2000              2000-12-31              325       1995-10-02   
3   2000              2000-12-31             1366       1992-05-11   
4   2000              2000-12-31              455       1992-08-01   

   ANNEE RECRUTEMENT DATE RETRAITE  ANNEE RETRAITE DATE PRISE SERVICE  \
0               1983    2011-02-13            2011         1983-10-01   
1               1992    2009-04-18            2009         1992-02-01   
2               1995    2013-03-08            2013         1995-10-02   
3               1992    2019-01-30            2019         1992-05-11   
4               1992    2030-08-28            2030         1992-08-01   

   ANNEE PRISE SERVICE DATE SORTIE  ... NBRE DE JOURS DE CONGE DU  \
0                 1983         NaT  ...                       NaN   
1                 1992         NaT  ...                       NaN   
2                 1995         NaT  ...                       NaN   
3                 1992         NaT  ...                       NaN   
4                 1992         NaT  ...                       NaN   

   ABSENCE DEDUCTIBLE CONGE  CLOTURE ETABLISSEMENT CONGE CLOTURE DRH CONGE  \
0                       NaN                          NaN               NaN   
1                       NaN                          NaN               NaN   
2                       NaN                          NaN               NaN   
3                       NaN                          NaN               NaN   
4                       NaN                          NaN               NaN   

  DATE OUVERTURE CONGE DRH CODE POSTE  \
0                      NaN  2111220.0   
1                      NaN  2110989.0   
2                      NaN  2111661.0   
3                      NaN  2110822.0   
4                      NaN  2111540.0   

                                      INTITULE POSTE  \
0              COMMIS D'ADMINISTRATION UNIVERSITAIRE   
1  SECRETAIRE STENOI-DACTYLOGRAPHE CORRESPONDANCI...   
2                                   AGENT DE SERVICE   
3                                          INGENIEUR   
4             EMPLOYÉ D'ADMINISTRATION UNIVERSITAIRE   

  DATE DEBUT OCCUPATION POSTE DATE FIN OCCUPATION POSTE NOMBRE  
0                  1983-10-01       2011-03-01 00:00:00      1  
1                  1997-05-01       2009-05-01 00:00:00      1  
2                  1995-10-02       2013-04-01 00:00:00      1  
3                  1992-05-11       2019-02-01 00:00:00      1  
4                  1992-07-01                       NaN      1  

[5 rows x 91 columns]

In [64]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1587 entries, 0 to 1586
Data columns (total 91 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   ANNEE                          1587 non-null   int64         
 1   DATE DE PRISE EN COMPTE        1587 non-null   datetime64[ns]
 2   NUMERO EMPLPOYE                1587 non-null   int64         
 3   DATE RECRUTEMENT               1587 non-null   datetime64[ns]
 4   ANNEE RECRUTEMENT              1587 non-null   int64         
 5   DATE RETRAITE                  1587 non-null   datetime64[ns]
 6   ANNEE RETRAITE                 1587 non-null   int64         
 7   DATE PRISE SERVICE             1587 non-null   datetime64[ns]
 8   ANNEE PRISE SERVICE            1587 non-null   int64         
 9   DATE SORTIE                    1 non-null      datetime64[ns]
 10  DATE TITULARISATION            19 non-null     datetime64[ns]
 11  ANNEE NAISSANCE  

In [65]:
data.columns

Index(['ANNEE', 'DATE DE PRISE EN COMPTE', 'NUMERO EMPLPOYE',
       'DATE RECRUTEMENT', 'ANNEE RECRUTEMENT', 'DATE RETRAITE',
       'ANNEE RETRAITE', 'DATE PRISE SERVICE', 'ANNEE PRISE SERVICE',
       'DATE SORTIE', 'DATE TITULARISATION', 'ANNEE NAISSANCE', 'AGE',
       'TRANCHE D'AGE', 'NATIONALITE', 'REGION DE NAISSANCE',
       'DEPARTEMENT DE NAISSANCE', 'CATEGORIE AGENT', 'SIGLE CATEGORIE AGENT',
       'TYPE AGENT', 'SIGLE TYPE AGENT', 'CIVILITE', 'SIGLE CIVILITE', 'SEXE',
       'DOCTORANT', 'GENRE', 'DISCIPLINE', 'LOGER', 'NOMBRE ENFANT',
       'NOMBRE EPOUSE', 'VEHICULER', 'STATUT AGENT', 'SIGLE STATUT',
       'DATE DEBUT STATUT', 'DATE FIN STATUT', 'POSITION', 'SIGLE POSITION',
       'DATE DEBUT POSITION', 'DATE FIN POSITION', 'SOUS POSITION',
       'SIGLE SOUS POSITION', 'CODE ETABLISSEMENT', 'ETABLISSEMENT',
       'SIGLE ETABLISSEMENT', 'DATE DEBUT AFFECTATION', 'DATE FIN AFFECTATION',
       'INTITULE FONCTION', 'DATE DEBUT DE FONCTION', 'DATE FIN DE FONCTION',
  

le fichier contient 91 colonnes et 1587 lignes 

#### anomalie1 : des personnes recrutes apres leur prise de services c'est a dire date_recrutement > data_prise de service 

In [66]:
def anomalie_recrutement(df):
    anomalie1_df = df[df['DATE RECRUTEMENT'] > (df['DATE PRISE SERVICE'])]
    if len(anomalie1_df) == 0:
        print("tous les agents ont une date de recrutement valide.")
    else:
     print(f"# {len(anomalie1_df)} agents ont une anomalie de date de recrutement. leur date de recrutement est supérieure à leur date de prise de service.")
     
anomalie_recrutement(data)     
    

# 5 agents ont une anomalie de date de recrutement. leur date de recrutement est supérieure à leur date de prise de service.


solution possible : supprimer ces lignes car leur faible nombre (5/1587) les rend negligeables 

#### anomalie2 : des personnes retraites avant d'etre recrutes c'est a dire date retraite < date recrutement 

In [67]:
def anomalie_retraite(df):
    anomalie1_df = df[df['DATE RETRAITE'] < (df['DATE RECRUTEMENT'])]
    if len(anomalie1_df) == 0:
        print("tous les agents ont une date de retraite valide (ceci n'est pas une anomalie !!!)")
    else:
     print(f"# {len(anomalie1_df)} agents ont une anomalie de date de retraite. leur date de retraite est inferieur à leur date de prise de recrutement.")
     
anomalie_retraite(data) 

tous les agents ont une date de retraite valide (ceci n'est pas une anomalie !!!)


#### anomalie3 : fin de poste avant le debut

In [68]:
def anomalie_fin_poste(df):
    #convert date columns to datetime
    df['DATE FIN DE FONCTION'] = pd.to_datetime(df['DATE FIN DE FONCTION'], errors='coerce')
    df['DATE DEBUT DE FONCTION'] = pd.to_datetime(df['DATE DEBUT DE FONCTION'], errors='coerce')
    # detect anomalies
    anomalie_df = df[df['DATE FIN DE FONCTION'] < df['DATE DEBUT DE FONCTION']]
    if len(anomalie_df) == 0:
        print("Tous les agents ont une date de fin de fonction valide (ceci n'est pas une anomalie !!!)")
    else:
        print(f"# {len(anomalie_df)} agents ont une anomalie de date de fin de fonction. Leur date de fin de fonction est inférieure à leur date de début de fonction.")
        
anomalie_fin_poste(data)        

Tous les agents ont une date de fin de fonction valide (ceci n'est pas une anomalie !!!)


#### anomalie4 : les valeurs manquantes dans chaque colonne

In [69]:
def valeur_manquante(df):
    anomalie_valeur = df.isnull().sum()
    anomalie_valeur = anomalie_valeur[anomalie_valeur > 0]
    if len(anomalie_valeur) == 0:
        print("Aucune valeur manquante dans le fichier.")
    else:
        print(f"# {len(anomalie_valeur)} colonnes contiennent des valeurs manquantes.")
        print(anomalie_valeur)
        
valeur_manquante(data)        

# 41 colonnes contiennent des valeurs manquantes.
DATE SORTIE                      1586
DATE TITULARISATION              1568
DEPARTEMENT DE NAISSANCE         1335
INTITULE FONCTION ANNEXE          448
DATE DEBUT DE FONCTION ANNEXE     449
DATE FIN DE FONCTION ANNEXE       449
DATE DEBUT GRADE                    1
DATE FIN GRADE                   1586
NOMBRE MOIS ANC                     1
DATE CREATION GRADE                 1
DATE PRISE EN COMPTE GRADE       1587
GRADE ACTIF                         1
INDICE                              1
CODE CORPS                          1
GRADE                               1
CODE TYPE GRADE AGENT               1
GRADE ASSOCIE                       1
NIVEAU STATUT                     664
TYPE AFFECTATION                 1250
TYPE SOUS STRUCTURE              1587
SOUS STRUCTURE                   1587
DATE DEBUT AFF SOUS STRUCTURE    1250
DATE FIN AFF SOUS STRUCTURE      1582
DATE DEBUT AFF STRUCTURE         1250
DATE FIN AFF STRUCTURE           1587


il y'a beaucoup de colonne avec des valeurs manquantes sur ce jeux de données

#### anomalie5 : detecter les doublons

In [70]:
#=====detecter les doublons dans le fichier========
def anomalie_doublons(df) : 
    doublons = df.duplicated().sum()
    numero_employe = df['NUMERO EMPLPOYE'].duplicated().sum()
    if doublons == 0:
        print("Aucun doublon dans le fichier.")
    else:
        print(f"# {doublons} doublons trouvés dans le fichier.")
        
    if numero_employe == 0: 
        print("Aucun numéro d'employé en double dans le fichier.")
    else:
        print(f"# {numero_employe} numéros d'employé en double trouvés dans le fichier.")    

anomalie_doublons(data) 

Aucun doublon dans le fichier.
# 1 numéros d'employé en double trouvés dans le fichier.


#### anomalie6 : les valeurs trop extemem comme les ages <18 ou > 65 les nb_enfants >10 ou nb_femmes > 4  

In [71]:
def anomalie6(data) :
    age = data['AGE']
    nb_enfants = data['NOMBRE ENFANT']
    femmes = data['NOMBRE EPOUSE']
    #age < 18 or age > 68
    anomalie_age = data[(age < 18) | (age > 68)]
    anomalie_nb_enfants = data[(nb_enfants < 0) | (nb_enfants > 10)]
    anomalie_nb_femmes = data[(femmes < 0) | (femmes > 4)]
    if len(anomalie_age) == 0:
        print("Aucun personnel n'a un age anormal.")
    else:
        print(f"# {len(anomalie_age)} agents ont un age anomale (soit inférieur à 18 ans soit supérieur à 68 ans).")
    if len(anomalie_nb_enfants) == 0:
        print("Aucun personnel n'a un nombre d'enfants anormal.")
    else:
        print(f"# {len(anomalie_nb_enfants)} agents ont un nombre d'enfants anomale (soit inférieur à 0 soit supérieur à 10).")  
    if len(anomalie_nb_femmes) == 0:
        print("Aucun personnel n'a un nombre d'épouses anormal.")
    else:
        print(f"# {len(anomalie_nb_femmes)} agents ont un nombre d'épouses anomale (soit inférieur à 0 soit supérieur à 4).")          


anomalie6(data) 

# 2 agents ont un age anomale (soit inférieur à 18 ans soit supérieur à 68 ans).
# 17 agents ont un nombre d'enfants anomale (soit inférieur à 0 soit supérieur à 10).
# 2 agents ont un nombre d'épouses anomale (soit inférieur à 0 soit supérieur à 4).


#### fonction pour detecter tous les anomalies a partir d'un fichier 

In [72]:
import os
import os
import io
import contextlib
import pandas as pd

def detecter_anomalies_personnel(fichier_entree ):
    # Lire le fichier (Excel ou CSV automatiquement détecté)
    extension = os.path.splitext(fichier_entree)[1].lower()
    if extension == '.csv':
        data = pd.read_csv(fichier_entree , encoding='latin1', sep=';')
    elif extension in ['.xls', '.xlsx']:
        data = pd.read_excel(fichier_entree)
    else:
        raise ValueError("Format de fichier non supporté. Utilisez .csv, .xls ou .xlsx")

    # Construire le nom du fichier de sortie
    nom_base = os.path.splitext(os.path.basename(fichier_entree))[0]
    nom_sortie = f"anomalies_{nom_base}.txt"

    # Liste des fonctions d'anomalies
    fonctions_anomalies = [anomalie_recrutement, anomalie_retraite, anomalie_fin_poste , anomalie_doublons , anomalie6]

    # Rediriger la sortie print vers un fichier texte
    with open(nom_sortie, 'w', encoding='utf-8') as f:
        for i, fonction in enumerate(fonctions_anomalies, 1):
            f.write(f"===== Anomalie {i} =====\n")
            buffer = io.StringIO()
            with contextlib.redirect_stdout(buffer):
                fonction(data)
            f.write(buffer.getvalue())
            f.write("\n\n")

    

    
